pkg/controller/replica.go

```diff
@@ -113,10 +114,47 @@ func (r *ReconcileReplica) Reconcile(ctx context.Context, request reconcile.Requ
 				}
 				// if replica attachment failed, delete the CRI and create a replacement
 			} else if azVolumeAttachment.Status.State == azdiskv1beta2.AttachmentFailed {
-				if err := r.cachedClient.Delete(ctx, azVolumeAttachment); err != nil {
-					return reconcile.Result{Requeue: true}, err
+				// if the failed attachment is retriable, reset the state to AttachmentPending
+				if azureutils.MapContains(azVolumeAttachment.Status.Annotations, consts.VolumeAttachRetryAnnotation) {
+					retryCount := 0
+					if value, exists := azureutils.GetFromMap(azVolumeAttachment.Status.Annotations, consts.VolumeAttachRetryCount); exists {
+						if n, err := strconv.Atoi(value); err == nil {
+							retryCount = n
+						} else {
+							r.logger.Error(err, "failed to convert retry count to int")
+						}
+					}
+
+					var updateFunc azureutils.UpdateCRIFunc
+					if retryCount < r.config.ControllerConfig.VolumeAttachRetryLimit {
+						updateFunc = func(obj client.Object) error {
+							azva := obj.(*azdiskv1beta2.AzVolumeAttachment)
+							azureutils.RemoveFromMap(azva.Status.Annotations, consts.VolumeAttachRetryAnnotation)
+							azureutils.AddToMap(azva.Status.Annotations, consts.VolumeAttachRetryCount, strconv.Itoa(retryCount+1))
+							_, derr := updateState(azva, azdiskv1beta2.AttachmentPending, forceUpdate)
+							azva.Status.Detail = nil
+							azva.Status.Error = nil
+							return derr
+						}
+					} else {
+						updateFunc = func(obj client.Object) error {
+							azva := obj.(*azdiskv1beta2.AzVolumeAttachment)
+							azureutils.AddToMap(azva.Status.Annotations, consts.VolumeDetachRequestAnnotation, "replica-controller")
+							_, derr := updateState(azva, azdiskv1beta2.AttachmentPending, forceUpdate)
+							return derr
+						}
+					}
+
+					if _, err := azureutils.UpdateCRIWithRetry(ctx, nil, r.cachedClient, r.azClient, azVolumeAttachment, updateFunc, consts.NormalUpdateMaxNetRetry, azureutils.UpdateCRIStatus); err != nil {
+						return reconcile.Result{Requeue: true}, err
+					}
+					// else, delete the CRI so that replace replica AzVolumeAttachment can be created.
+				} else {
+					if err := r.cachedClient.Delete(ctx, azVolumeAttachment); err != nil {
```

Comment:
I find it easier to see the flow of code if the smaller of the `if..else` clauses is first, e.g.

```go
if !azureutils.MapContains(...) {
    // delete AzVolumeAttachment CRI clause
} else {
    // much larger retry handling clause}
```

This is especially true when one clause is *MUCH* larger than the other.

Reply:
Nice suggestion! Thank you.